src/coreComponents/constitutive/fluid/multifluid/compositional/functions/StabilityTest.hpp

```diff
@@ -0,0 +1,231 @@
+                                                  logFugacity );
+        real64 error = 0.0;
+        for( integer const ic : presentComponents )
+        {
+          real64 const dG =  logTrialComposition[ic] + logFugacity[ic] - hyperplane[ic];
+          error += (dG*dG);
+        }
+        error = LvArray::math::sqrt( error );
+
+        if( error < MultiFluidConstants::fugacityTolerance )
+        {
+          // Calculate modified tangent plane distance (Michelsen, 1982b) of trial composition relative to input composition
+          double tpd = 1.0;
+          for( integer const ic : presentComponents )
+          {
+            tpd += trialComposition( trialIndex, ic ) * (logTrialComposition[ic] + logFugacity[ic] - hyperplane[ic] - 1.0);
+          }
+          if( tpd < tangentPlaneDistance )
+          {
+            tangentPlaneDistance = tpd;
+          }
+          numberOfStationaryPoints++;
+          break;
+        }
+      }
+    }
+    if( numberOfStationaryPoints == numTrials )
+    {
+      for( integer const ic : presentComponents )
+      {
+        kValues[ic] = trialComposition( 1, ic ) / trialComposition( 0, ic );
+      }
+    }
+    return numberOfStationaryPoints == numTrials;
+  }
+
+private:
+  /**
+   * @brief Calculate which components are present.
+   * @details Creates a list of indices whose components have non-zero mole fraction.
+   * @param[in] numComps number of components
+   * @param[in] composition the composition of the fluid
+   * @param[out] presentComponents the list of present components
+   * @return the number of present components
+   */
+  template< typename ARRAY >
+  GEOS_HOST_DEVICE
+  GEOS_FORCE_INLINE
+  static integer calculatePresentComponents( integer const numComps,
+                                             arraySlice1d< real64 const > const & composition,
+                                             ARRAY & presentComponents )
+  {
+    // Check for machine-zero feed values
+    integer presentCount = 0;
+    for( integer ic = 0; ic < numComps; ++ic )
+    {
+      if( MultiFluidConstants::epsilon < composition[ic] )
+      {
+        presentComponents[presentCount++] = ic;
+      }
+    }
+    presentComponents.resize( presentCount );
+    return presentCount;
+  }
+
+  /**
+   * @brief Normalise a composition in place to ensure that the components add up to unity
+   * @param[in] numComps number of components
+   * @param[in/out] composition composition to be normalized
+   * @return the sum of the given values
+   */
+  template< integer USD >
+  GEOS_HOST_DEVICE
+  GEOS_FORCE_INLINE
+  static real64 normalizeComposition( integer const numComps,
+                                      arraySlice1d< real64, USD > const & composition )
+  {
+    real64 totalMoles = 0.0;
+    for( integer ic = 0; ic < numComps; ++ic )
+    {
+      totalMoles += composition[ic];
+    }
+    real64 const oneOverTotalMoles = 1.0 / (totalMoles + MultiFluidConstants::epsilon);
```

Comment:
wouldn't it be safer to throw/return 1 on null `totalMoles` ?

Reply:
I have added an assert.